hamilton/driver.py

```diff
@@ -657,24 +677,55 @@ def raw_execute(
         inputs: Dict[str, Any] = None,
         _fn_graph: graph.FunctionGraph = None,
     ) -> Dict[str, Any]:
-        """Raw execute function that does the meat of execute.
-
-        Don't use this entry point for execution directly. Always go through `.execute()`.
+        """Don't use this entry point for execution directly. Always go through `.execute()`.
         In case you are using `.raw_execute()` directly, please switch to `.execute()` using a
         `base.DictResult()`. Note: `base.DictResult()` is the default return of execute if you are
         using the `driver.Builder()` class to create a `Driver()` object.
+        """
+        success = True
+        error = None
+        results = None
+        try:
+            return self.__raw_execute(final_vars, overrides, display_graph, inputs=inputs)
+        except Exception as e:
+            success = False
+            logger.error(SLACK_ERROR_MESSAGE)
+            error = e
+            raise e
+        finally:
+            if self.adapter.does_hook("post_graph_execute", is_async=False):
+                self.adapter.call_all_lifecycle_hooks_sync(
+                    "post_graph_execute",
+                    run_id=self.run_id,
+                    graph=self.function_graph,
+                    success=success,
+                    error=error,
+                    results=results,
+                )
+
+    def __raw_execute(
+        self,
+        final_vars: List[str],
+        overrides: Dict[str, Any] = None,
+        display_graph: bool = False,
+        inputs: Dict[str, Any] = None,
+        _fn_graph: graph.FunctionGraph = None,
+    ) -> Dict[str, Any]:
+        """Raw execute function that does the meat of execute.
+
+        Private method since the result building and post_graph_execute lifecycle hooks are performed outside and so this returns an incomplete result.
 
         :param final_vars: Final variables to compute
         :param overrides: Overrides to run.
         :param display_graph: DEPRECATED. DO NOT USE. Whether or not to display the graph when running it
         :param inputs: Runtime inputs to the DAG
         :return:
         """
-        function_graph = _fn_graph if _fn_graph is not None else self.graph
-        run_id = str(uuid.uuid4())
-        nodes, user_nodes = function_graph.get_upstream_nodes(final_vars, inputs, overrides)
+        self.function_graph = _fn_graph if _fn_graph is not None else self.graph
+        self.run_id = str(uuid.uuid4())
+        nodes, user_nodes = self.function_graph.get_upstream_nodes(final_vars, inputs, overrides)
```

Comment:
I don't like this structure as is.. the pre-execute hooks are in one function and the post-execute hooks are outside. The ```self``` was a workaround to get the test to pass and not running into issues with unspecified ```run_id``` when an exception occurs.

Maybe we can simplify this and just put the result_builder into __raw_execute? This would make everything neater, planning to code this up tonight.

Reply:
Let me know if this solution is acceptable or if it breaks something somewhere else?

Reply:
So yeah, results builder is not in raw execute. The problem here is that it's not thread-safe -- this is something we rely on (E.G. being able to have multiple calls to the driver at once). So I don't think this will work?

IMO it's OK having it in two different places, but I think there are better solution here:
1. (simple, hackyish) Have `__raw_execute` return the run_id/function graph -- function graph can actually always be passed in.
2. Cleaner -- pass run_id into `__raw_execute` -- why should it generate? Then if we pass that + the graph in we can easily call it at the beginning. 
3. Probably best -- just move some of the logic outside of `__raw_execute`, always surrounding it by the pre/post graph hooks. Then it look like a sandwich, but has the right tooling. The `raw_execute` function is the only one that does not call the result builder, the rest call the post graph execute hook after calling the result builder.

One thing to note -- the earlier the pre graph execute and later the post graph execute hooks are called the better. This is because it can capture errors -- thus any configuration problems/hamilton bugs will be caught in the right place and displayed in the UI.

Reply:
I was 90% sure my suggestion will not work; looked too easy :) I adjusted the ```.materialize()``` entyr point in the same way since it also calls ```__raw_execute``` and all the test pass now as well.

Reply:
Nice, thanks! Yeah we should have tests for this, but it's always an issue of how much you can reasonably cover...